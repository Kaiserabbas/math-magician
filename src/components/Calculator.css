.calculator {
  width: 300px;
  background-color: #f4f4f4;
  border: 1px solid #ccc;
  border-radius: 8px;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
  margin: 100px auto 0 auto;
  padding: 20px;
}

.display {
  height: 30px;
  background-color: #222;
  color: #fff;
  font-size: 24px;
  text-align: right;
  padding: 10px;
  border-radius: 6px;
  margin-bottom: 10px;
}

.buttons {
  display: grid;
  grid-template-rows: repeat(4, 1fr);
    grid-gap: 10px;
}

.button {
  width: 100%;
  padding: 10px;
    border-radius: 6px;
  font-size: 14px;
  border: 1px solid #ccc;
  background-color: #e0e0e0;
  cursor: pointer;
    transition: background-color 0.3s;
}

.row {
  display: flex;
  gap: 10px;
}

.button:hover {
  background-color: #ccc;
}

.operator {
  background-color: #f49313dd;
  color: white;
  font-weight: bold;
}

.row5 {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
}

.row5 :first-child {
  grid-column: span 2;
}

.quote {
  color: #9c5b05;
  position: relative;
  width: 300px;
  font-style: italic;
  font-size: 18px;
  text-align: center;
  margin: 20px auto;}
