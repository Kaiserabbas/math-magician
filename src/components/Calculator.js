import React, { useState } from 'react';
import './Calculator.css';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [calculatorData, setCalculatorData] = useState({});
  const handleButtonClick = (buttonName) => {
    console.log('Button clicked:', buttonName); // Add this line
    if (buttonName === '=') {
      const result = calculate(calculatorData, '=');
      setCalculatorData(result);
    } else {
      const newData = calculate(calculatorData, buttonName);
      console.log('New data:', newData);
      setCalculatorData(newData);
    }
  };

  const renderButton = (label, ClassName) => {
    return (
      <button
        type="button"
        className={ClassName}
        onClick={() => handleButtonClick(label)}
      >
        {label}
      </button>
    );
  };

  return (
    <div className="calculator">
      <div className="display">
        {' '}
        {calculatorData.total || calculatorData.next || '0'}
      </div>
      <div className="buttons">
        <div className="row1 row">
          {renderButton('AC', 'button')}
          {renderButton('+/-', 'button')}
          {renderButton('%', 'button')}
          {renderButton('รท', ' button operator')}
        </div>
        <div className="row2 row">
          {renderButton('7', 'button')}
          {renderButton('8', 'button')}
          {renderButton('9', 'button')}
          {renderButton('x', 'button operator')}
        </div>
        <div className="row3 row">
          {renderButton('4', 'button')}
          {renderButton('5', 'button')}
          {renderButton('6', 'button')}
          {renderButton('-', 'button operator')}
        </div>
        <div className="row4 row">
          {renderButton('1', 'button')}
          {renderButton('2', 'button')}
          {renderButton('3', 'button')}
          {renderButton('+', 'button operator')}
        </div>
        <div className="row5 row">
          {renderButton('0', 'button')}
          {renderButton('.', 'button')}
          {renderButton('=', 'button operator')}
        </div>
      </div>
    </div>
  );
};

export default Calculator;
