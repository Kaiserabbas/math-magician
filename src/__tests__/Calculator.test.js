// Calculator.test.js
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Calculator from '../components/Calculator';

describe('Calculator component', () => {
  it('renders the component correctly', () => {
    render(<Calculator />);
    expect(screen.getByText("Let's do some math")).toBeInTheDocument();
  });

  it('handles number button clicks and updates display', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('3'));

    expect(display).toHaveTextContent('123');
  });

  it('handles operation button clicks and updates display', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('='));

    expect(display).toHaveTextContent('8');
  });

  it('handles decimal button click and updates display', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('5'));

    expect(display).toHaveTextContent('1.5');
  });

  it('handles AC button click and resets display', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('7'));
    fireEvent.click(screen.getByText('AC'));

    expect(display).toHaveTextContent('0');
  });
  it('handles +/- button click and toggles sign', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('9'));
    fireEvent.click(screen.getByText('+/-'));

    expect(display).toHaveTextContent('-9');
  });

  it('handles % button click and calculates percentage', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('8'));
    fireEvent.click(screen.getByText('%'));

    expect(display).toHaveTextContent('0.08');
  });

  it('handles multiple operations correctly', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');

    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('-'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('='));

    expect(display).toHaveTextContent('6');
  });

  it('matches the snapshot', () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });
});
