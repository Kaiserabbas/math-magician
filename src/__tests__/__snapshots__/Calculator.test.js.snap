// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component matches the snapshot 1`] = `
<div>
  <div
    class="calculator-container"
  >
    <div
      class="calculator-text"
    >
      <h2>
        Let's do some math
      </h2>
    </div>
    <div
      class="calculator"
    >
      <div
        class="display"
        data-testid="display"
      />
      <div
        class="buttons"
      >
        <div
          class="row1 row"
        >
          <button
            class="button"
            type="button"
          >
            AC
          </button>
          <button
            class="button"
            type="button"
          >
            +/-
          </button>
          <button
            class="button"
            type="button"
          >
            %
          </button>
          <button
            class=" button operator"
            type="button"
          >
            รท
          </button>
        </div>
        <div
          class="row2 row"
        >
          <button
            class="button"
            type="button"
          >
            7
          </button>
          <button
            class="button"
            type="button"
          >
            8
          </button>
          <button
            class="button"
            type="button"
          >
            9
          </button>
          <button
            class="button operator"
            type="button"
          >
            x
          </button>
        </div>
        <div
          class="row3 row"
        >
          <button
            class="button"
            type="button"
          >
            4
          </button>
          <button
            class="button"
            type="button"
          >
            5
          </button>
          <button
            class="button"
            type="button"
          >
            6
          </button>
          <button
            class="button operator"
            type="button"
          >
            -
          </button>
        </div>
        <div
          class="row4 row"
        >
          <button
            class="button"
            type="button"
          >
            1
          </button>
          <button
            class="button"
            type="button"
          >
            2
          </button>
          <button
            class="button"
            type="button"
          >
            3
          </button>
          <button
            class="button operator"
            type="button"
          >
            +
          </button>
        </div>
        <div
          class="row5 row"
        >
          <button
            class="button"
            type="button"
          >
            0
          </button>
          <button
            class="button"
            type="button"
          >
            .
          </button>
          <button
            class="button operator"
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
